File: mac/FluentCore/Sources/FluentCore/GeminiClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Transcription through the unary Interactions endpoint with `gemini-3.5-transcribe`, the same
/// request the Android app sends as its fallback. The iOS app records a whole turn and sends it
/// in one request, which suits the keyboard handoff: nothing streams while the user is switching
/// between apps.
public struct GeminiClient: Sendable {
    public let apiKey: String
    public let session: URLSession
    /// The Interactions endpoint. Only tests point it anywhere else (a local mock server).
    public let endpoint: URL
    public let modelsBase: URL

    public init(apiKey: String, session: URLSession = .shared,
                endpoint: URL = Constants.interactionsURL, modelsBase: URL = Constants.modelsURL) {
        self.apiKey = apiKey
        self.session = session
        self.endpoint = endpoint
        self.modelsBase = modelsBase
    }

    public func transcribe(
        wav: Data,
        mode: TranscriptionMode,
        languageCodes: [String],
        vocabulary: [String]
    ) async -> Result<String, TranscriptionError> {
        guard !apiKey.isEmpty else { return .failure(.noApiKey) }
        let request = Self.transcribeRequest(apiKey: apiKey, wav: wav, mode: mode,
                                             languageCodes: languageCodes, vocabulary: vocabulary,
                                             endpoint: endpoint)
        do {
            let (data, response) = try await session.data(for: request)
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard (200..<300).contains(code) else { return .failure(Self.mapHTTPError(code: code, body: data)) }
            guard let text = Self.extractText(data)?.trimmingCharacters(in: .whitespacesAndNewlines),
                  !text.isEmpty else { return .failure(.noSpeech) }
            return .success(text)
        } catch {
            return .failure(Self.mapTransportError(error))
        }
    }

    /// Validates a key cheaply by fetching the model; used by Settings → Test connection.
    public func testConnection() async -> Result<String, TranscriptionError> {
        guard !apiKey.isEmpty else { return .failure(.noApiKey) }
        var request = URLRequest(url: modelsBase.appendingPathComponent(Constants.batchModel))
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
        request.timeoutInterval = 20
        do {
            let (data, response) = try await session.data(for: request)
            let code = (response as? HTTPURLResponse)?.statusCode ?? 0
            return (200..<300).contains(code) ? .success(Constants.batchModel)
                : .failure(Self.mapHTTPError(code: code, body: data))
        } catch {
            return .failure(Self.mapTransportError(error))
        }
    }

    // MARK: request

    public static func transcribeRequest(
        apiKey: String, wav: Data, mode: TranscriptionMode,
        languageCodes: [String], vocabulary: [String],
        endpoint: URL = Constants.interactionsURL
    ) -> URLRequest {
        var config: [String: Any] = [
            "mode": mode == .smart ? "smart" as Any : ["type": "verbatim"] as Any,
            "language_codes": languageCodes,
        ]
        if !vocabulary.isEmpty { config["custom_vocabulary"] = Array(vocabulary.prefix(1000)) }
        let body: [String: Any] = [
            "model": Constants.batchModel,
            "input": [["type": "audio", "data": wav.base64EncodedString(), "mime_type": "audio/wav"]],
            "generation_config": ["transcription_config": config],
        ]
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-goog-api-key")
        request.setValue(Constants.apiRevision, forHTTPHeaderField: "Api-Revision")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body, options: [.sortedKeys])
        request.timeoutInterval = Constants.batchTimeout
        return request
    }

    // MARK: response

    /// The Interactions response carries the transcript as text content inside `steps`.
    /// Falls back to `output_text` and to the classic `candidates` shape so a server-side
    /// response-format change cannot silently drop a transcript.
    public static func extractText(_ data: Data) -> String? {
        guard let root = rootObject(data) else { return nil }

        if let text = root["output_text"] as? String, !text.trimmingCharacters(in: .whitespaces).isEmpty {
            return text
        }
        if let steps = root["steps"] as? [[String: Any]] {
            let texts = steps.flatMap { ($0["content"] as? [[String: Any]]) ?? [] }
                .filter { ($0["type"] as? String) == "text" }
                .compactMap { $0["text"] as? String }
            let joined = texts.joined(separator: " ")
            if !joined.trimmingCharacters(in: .whitespaces).isEmpty { return joined }
        }
        if let candidates = root["candidates"] as? [[String: Any]],
           let parts = (candidates.first?["content"] as? [String: Any])?["parts"] as? [[String: Any]] {
            let joined = parts.compactMap { $0["text"] as? String }.joined()
            if !joined.trimmingCharacters(in: .whitespaces).isEmpty { return joined }
        }
        return nil
    }

    public static func mapHTTPError(code: Int, body: Data) -> TranscriptionError {
        let error = rootObject(body)?["error"] as? [String: Any]
        let message = (error?["message"] as? String) ?? ""
        let status = (error?["status"] as? String) ?? ""
        let reasons = ((error?["details"] as? [[String: Any]]) ?? []).compactMap { $0["reason"] as? String }

        let keyInvalid = reasons.contains { $0.localizedCaseInsensitiveContains("API_KEY") }
            || message.localizedCaseInsensitiveContains("API key not valid")
            || message.localizedCaseInsensitiveContains("API_KEY_INVALID")
        let quota = code == 429 || status == "RESOURCE_EXHAUSTED"
            || message.localizedCaseInsensitiveContains("quota")
            || message.localizedCaseInsensitiveContains("rate limit")

        if keyInvalid { return .invalidApiKey }
        if quota { return .quotaExceeded }
        if code == 401 || code == 403 || status == "PERMISSION_DENIED" { return .invalidApiKey }
        return .server(code: code, detail: message)
    }

    public static func mapTransportError(_ error: Error) -> TranscriptionError {
        guard let urlError = error as? URLError else { return .unknown(error.localizedDescription) }
        switch urlError.code {
        case .notConnectedToInternet, .cannotFindHost, .dnsLookupFailed, .dataNotAllowed,
             .internationalRoamingOff, .cannotConnectToHost:
            return .noNetwork
        case .timedOut, .networkConnectionLost:
            return .connectionLost
        default:
            return .unknown(urlError.localizedDescription)
        }
    }

    /// Bodies come back either as a bare object or wrapped in a single-element array depending on
    /// the endpoint, so both shapes are unwrapped here.
    private static func rootObject(_ data: Data) -> [String: Any]? {
        let parsed = try? JSONSerialization.jsonObject(with: data)
        if let object = parsed as? [String: Any] { return object }
        return (parsed as? [Any])?.first as? [String: Any]
    }
}

File: src/Fluent.Core/GeminiClient.cs
using System.Net;
using System.Net.Http.Headers;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Fluent.Core;

/// <summary>Transcription through the unary Interactions endpoint with <c>gemini-3.5-transcribe</c>, the
/// same request the Android and Mac apps send as their fallback.</summary>
public sealed class GeminiClient(string apiKey, HttpClient? http = null, Uri? endpoint = null, Uri? modelsBase = null)
{
    static readonly HttpClient Shared = new() { Timeout = Timeout.InfiniteTimeSpan };
    readonly HttpClient http = http ?? Shared;
    readonly Uri endpoint = endpoint ?? Constants.InteractionsUrl;
    readonly Uri modelsBase = modelsBase ?? Constants.ModelsUrl;

    public async Task<TranscriptionResult> TranscribeAsync(byte[] wav, TranscriptionMode mode,
        IReadOnlyList<string> languageCodes, IReadOnlyList<string> vocabulary, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(apiKey)) return TranscriptionResult.Fail(TranscriptionError.NoApiKey);
        using var request = TranscribeRequest(apiKey, wav, mode, languageCodes, vocabulary, endpoint);
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(Constants.BatchTimeout);
        try
        {
            using var response = await http.SendAsync(request, timeout.Token).ConfigureAwait(false);
            var body = await response.Content.ReadAsByteArrayAsync(timeout.Token).ConfigureAwait(false);
            var code = (int)response.StatusCode;
            if (code is < 200 or >= 300) return TranscriptionResult.Fail(MapHttpError(code, body));
            var text = ExtractText(body)?.Trim();
            return string.IsNullOrEmpty(text) ? TranscriptionResult.Fail(TranscriptionError.NoSpeech) : TranscriptionResult.Ok(text);
        }
        catch (Exception e) when (!ct.IsCancellationRequested)
        {
            return TranscriptionResult.Fail(MapTransportError(e));
        }
    }

    /// <summary>Validates a key cheaply by fetching the model; used by Settings → Test connection.</summary>
    public async Task<TranscriptionResult> TestConnectionAsync(CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(apiKey)) return TranscriptionResult.Fail(TranscriptionError.NoApiKey);
        using var request = new HttpRequestMessage(HttpMethod.Get, new Uri(modelsBase, Constants.BatchModel));
        request.Headers.Add("x-goog-api-key", apiKey);
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(ct);
        timeout.CancelAfter(TimeSpan.FromSeconds(20));
        try
        {
            using var response = await http.SendAsync(request, timeout.Token).ConfigureAwait(false);
            var body = await response.Content.ReadAsByteArrayAsync(timeout.Token).ConfigureAwait(false);
            var code = (int)response.StatusCode;
            return code is >= 200 and < 300 ? TranscriptionResult.Ok(Constants.BatchModel)
                : TranscriptionResult.Fail(MapHttpError(code, body));
        }
        catch (Exception e) when (!ct.IsCancellationRequested)
        {
            return TranscriptionResult.Fail(MapTransportError(e));
        }
    }

    public static HttpRequestMessage TranscribeRequest(string apiKey, byte[] wav, TranscriptionMode mode,
        IReadOnlyList<string> languageCodes, IReadOnlyList<string> vocabulary, Uri? endpoint = null)
    {
        var config = new JsonObject
        {
            ["language_codes"] = new JsonArray(languageCodes.Select(c => (JsonNode)JsonValue.Create(c)!).ToArray()),
            ["mode"] = mode == TranscriptionMode.Smart ? JsonValue.Create("smart") : new JsonObject { ["type"] = "verbatim" },
        };
        if (vocabulary.Count > 0)
            config["custom_vocabulary"] = new JsonArray(vocabulary.Take(1000).Select(v => (JsonNode)JsonValue.Create(v)!).ToArray());
        var body = new JsonObject
        {
            ["generation_config"] = new JsonObject { ["transcription_config"] = config },
            ["input"] = new JsonArray(new JsonObject
            {
                ["data"] = Convert.ToBase64String(wav), ["mime_type"] = "audio/wav", ["type"] = "audio",
            }),
            ["model"] = Constants.BatchModel,
        };
        var request = new HttpRequestMessage(HttpMethod.Post, endpoint ?? Constants.InteractionsUrl)
        {
            Content = new StringContent(body.ToJsonString(), Encoding.UTF8, "application/json"),
        };
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
        request.Headers.Add("x-goog-api-key", apiKey);
        request.Headers.Add("Api-Revision", Constants.ApiRevision);
        return request;
    }

    /// <summary>The Interactions response carries the transcript as text content inside <c>steps</c>.
    /// Falls back to <c>output_text</c> and to the classic <c>candidates</c> shape.</summary>
    public static string? ExtractText(byte[] data)
    {
        var root = RootObject(data);
        if (root is null) return null;
        if (root["output_text"] is JsonValue ov && ov.TryGetValue<string>(out var output) && !string.IsNullOrWhiteSpace(output))
            return output;
        if (root["steps"] is JsonArray steps)
        {
            var texts = steps.OfType<JsonObject>()
                .SelectMany(s => s["content"] as JsonArray ?? [])
                .OfType<JsonObject>()
                .Where(c => Str(c["type"]) == "text")
                .Select(c => Str(c["text"]))
                .Where(t => t is not null);
            var joined = string.Join(" ", texts);
            if (!string.IsNullOrWhiteSpace(joined)) return joined;
        }
        if (root["candidates"] is JsonArray cands && cands.FirstOrDefault() is JsonObject first
            && first["content"] is JsonObject content && content["parts"] is JsonArray parts)
        {
            var joined = string.Concat(parts.OfType<JsonObject>().Select(p => Str(p["text"]) ?? ""));
            if (!string.IsNullOrWhiteSpace(joined)) return joined;
        }
        return null;
    }

    public static TranscriptionError MapHttpError(int code, byte[] body)
    {
        var error = RootObject(body)?["error"] as JsonObject;
        var message = Str(error?["message"]) ?? "";
        var status = Str(error?["status"]) ?? "";
        var reasons = (error?["details"] as JsonArray ?? []).OfType<JsonObject>().Select(d => Str(d["reason"]) ?? "");

        var keyInvalid = reasons.Any(r => r.Contains("API_KEY", StringComparison.OrdinalIgnoreCase))
            || message.Contains("API key not valid", StringComparison.OrdinalIgnoreCase)
            || message.Contains("API_KEY_INVALID", StringComparison.OrdinalIgnoreCase);
        var quota = code == 429 || status == "RESOURCE_EXHAUSTED"
            || message.Contains("quota", StringComparison.OrdinalIgnoreCase)
            || message.Contains("rate limit", StringComparison.OrdinalIgnoreCase);

        if (keyInvalid) return TranscriptionError.InvalidApiKey;
        if (quota) return TranscriptionError.QuotaExceeded;
        if (code is 401 or 403 || status == "PERMISSION_DENIED") return TranscriptionError.InvalidApiKey;
        return TranscriptionError.Server(code, message);
    }

    public static TranscriptionError MapTransportError(Exception e) => e switch
    {
        TaskCanceledException or OperationCanceledException or TimeoutException => TranscriptionError.ConnectionLost,
        HttpRequestException { InnerException: SocketException se } when se.SocketErrorCode is
            SocketError.HostNotFound or SocketError.NoData or SocketError.NetworkUnreachable
            or SocketError.HostUnreachable or SocketError.ConnectionRefused or SocketError.TryAgain => TranscriptionError.NoNetwork,
        HttpRequestException { HttpRequestError: HttpRequestError.NameResolutionError or HttpRequestError.ConnectionError } => TranscriptionError.NoNetwork,
        HttpRequestException { InnerException: IOException } => TranscriptionError.ConnectionLost,
        _ => TranscriptionError.Unknown(e.Message),
    };

    internal static string? Str(JsonNode? n) => n is JsonValue v && v.TryGetValue<string>(out var s) ? s : null;

    /// <summary>Bodies come back either as a bare object or wrapped in a single-element array.</summary>
    static JsonObject? RootObject(byte[] data)
    {
        try
        {
            var parsed = JsonNode.Parse(data);
            return parsed as JsonObject ?? (parsed as JsonArray)?.FirstOrDefault() as JsonObject;
        }
        catch { return null; }
    }
}
